flink-python/src/main/java/org/apache/flink/table/runtime/operators/python/aggregate/arrow/stream/AbstractStreamArrowPythonBoundedRowsOperator.java

```diff
@@ -0,0 +1,257 @@
+			RowData data = arrowSerializer.read(i);
+			RowData key = forwardedInputQueue.poll();
+			reuseJoinedRow.setRowKind(key.getRowKind());
+			rowDataWrapper.collect(reuseJoinedRow.replace(key, data));
+		}
+	}
+
+	void registerProcessingCleanupTimer(long currentTime) throws Exception {
+		if (stateCleaningEnabled) {
+			registerProcessingCleanupTimer(
+				cleanupTsState,
+				currentTime,
+				minRetentionTime,
+				maxRetentionTime,
+				timerService);
+		}
+	}
+
+	void triggerWindowProcess(List<RowData> inputs, int i, int index) throws Exception {
+		int startIndex;
+		int startPos = 0;
+		if (windowData.isEmpty()) {
+			if (i >= lowerBoundary) {
+				for (int j = (int) (i - lowerBoundary); j <= i; j++) {
+					RowData rowData = inputs.get(j);
+					windowData.add(rowData);
+					arrowSerializer.write(getFunctionInput(rowData));
+				}
+				currentBatchCount += lowerBoundary;
+			} else {
+				Long previousTimestamp;
+				List<RowData> previousData = null;
+				int length = 0;
+				startIndex = index - 1;
+				long remainingDataCount = lowerBoundary - i;
+				ListIterator<Long> iter = sortedTimestamps.listIterator(index);
+				while (remainingDataCount > 0 && iter.hasPrevious()) {
+					previousTimestamp = iter.previous();
+					previousData = inputState.get(previousTimestamp);
+					length = previousData.size();
+					if (remainingDataCount <= length) {
+						startPos = (int) (length - remainingDataCount);
+						remainingDataCount = 0;
+					} else {
+						remainingDataCount -= length;
+						startIndex--;
+					}
+				}
+				if (previousData != null) {
+					for (int j = startPos; j < length; j++) {
+						RowData rowData = previousData.get(j);
+						windowData.add(rowData);
+						arrowSerializer.write(getFunctionInput(rowData));
+						currentBatchCount++;
+					}
+					// clear outdated data.
+					ListIterator<Long> clearIter = sortedTimestamps.listIterator();
+					for (int j = 0; j < startIndex; j++) {
+						long outdatedTs = clearIter.next();
+						inputState.remove(outdatedTs);
+					}
+
+					startIndex++;
+					while (startIndex < index) {
+						previousTimestamp = iter.next();
+						previousData = inputState.get(previousTimestamp);
+						for (RowData rowData : previousData) {
+							windowData.add(rowData);
+							arrowSerializer.write(getFunctionInput(rowData));
+						}
+						currentBatchCount += previousData.size();
+						startIndex++;
+					}
+				}
+				for (int j = 0; j <= i; j++) {
+					RowData rowData = inputs.get(j);
+					windowData.add(rowData);
+					arrowSerializer.write(getFunctionInput(rowData));
+				}
+				currentBatchCount += i + 1;
+			}
+		} else {
+			if (windowData.size() <= lowerBoundary) {
+				windowData.add(inputs.get(i));
+			} else {
+				windowData.pop();
+				windowData.add(inputs.get(i));
+			}
+			for (RowData rowData : windowData) {
+				arrowSerializer.write(getFunctionInput(rowData));
+			}
+			currentBatchCount += windowData.size();
+		}
+		if (currentBatchCount > 0) {
```

Comment:
Extract these code into a separate class and share them between rows and ranges over window?